entities/ctxlock/ctx_lock.go

```diff
@@ -0,0 +1,194 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package ctxlock
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	"github.com/weaviate/weaviate/usecases/monitoring"
+)
+
+var ErrCtxTimeout = errors.New("ctxsync: lock acquisition timed out")
+
+type CtxRWMutex struct {
+	location string
+
+	readerCount int32         // number of active readers
+	writer      chan struct{} // single-slot semaphore for writers
+	mu          sync.Mutex    // protects coordination
+}
+
+func NewCtxRWMutex(location string) *CtxRWMutex {
+	return &CtxRWMutex{
+		location: location,
+		writer:   make(chan struct{}, 1),
+	}
+}
+
+func (m *CtxRWMutex) CtxRWLocation(location string) {
+	m.location = location
+}
+
+//
+// Write locks
+//
+
+func (m *CtxRWMutex) Lock() {
+	monitoring.GetMetrics().LocksWaiting.WithLabelValues(m.location).Inc()
+	m.writer <- struct{}{}
+	m.waitForReaders()
+	monitoring.GetMetrics().LocksWaiting.WithLabelValues(m.location).Dec()
+	monitoring.GetMetrics().Locks.WithLabelValues(m.location).Inc()
+}
+
+func (m *CtxRWMutex) Unlock() {
+	select {
+	case <-m.writer:
+		monitoring.GetMetrics().Locks.WithLabelValues(m.location).Dec()
+	default:
+		panic("unlock called without a matching lock")
+	}
+}
+
+func (m *CtxRWMutex) LockContext(ctx context.Context) error {
+	monitoring.GetMetrics().LocksWaiting.WithLabelValues(m.location).Inc()
+	select {
+	case m.writer <- struct{}{}:
+		// Got writer slot
+	case <-ctx.Done():
+		monitoring.GetMetrics().LocksWaiting.WithLabelValues(m.location).Dec()
+		return ErrCtxTimeout
+	}
+	m.waitForReaders()
+	monitoring.GetMetrics().LocksWaiting.WithLabelValues(m.location).Dec()
+	monitoring.GetMetrics().Locks.WithLabelValues(m.location).Inc()
+	return nil
+}
+
+//
+// Read locks
+//
+
+func (m *CtxRWMutex) RLock() {
+	monitoring.GetMetrics().LocksWaiting.WithLabelValues(m.location).Inc()
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// If no writer currently holds the lock, increment readers
+	if len(m.writer) == 0 {
+		atomic.AddInt32(&m.readerCount, 1)
+		monitoring.GetMetrics().LocksWaiting.WithLabelValues(m.location).Dec()
+		monitoring.GetMetrics().Locks.WithLabelValues(m.location).Inc()
+		return
+	}
+
+	// Wait until writer is gone
+	for len(m.writer) > 0 {
+		m.mu.Unlock()
+		m.mu.Lock()
```

Comment:
How does this part work?

My understanding is that it is a way for to let other goroutines calling `Rlock` to proceed? Seeing how the method acquires a lock in the beginning. Is that correct?

I don't quite understand what `m.mu` guards -- the `readerCount` is incremented atomically.